### ISSUE TITLE
Needless / unnecessary Result

### ISSUE THREAD

**Issue by emilk** [rust-lang/rust-clippy#6265]
### What it does

``` rust
fn foo() -> Result<i32, Error> {
    Ok(42)
}
```

Suggested clippy lint: "Needless use of `Result` (function never returns `Err`). Consider changing the return type to `i32` ."

(Unsure if it should be "needless" or "unnecessary", see https://github.com/rust-lang/rust-clippy/issues/2845)

### Categories (optional)

`clippy::complexity` (code that does something simple but in a complex way)

The use of this lint would most often come up after refactoring some code, when a function that used to have a `?` in it no longer thus, and so now never returns an error.

Removing unnecessary uses of `Result` makes your code cleaner and clearer.

### Drawbacks

Sometimes the user wants to return a `Result` because they plan on adding a failure condition later.

### Example

```rust
fn foo() -> Result<i32, Error> {
    Ok(42)
}
```

Could be written as:

```rust
fn foo() -> i32 {
    42
}
```

**Comment by llogiq** [comment]
Let's only lint this on non-exported `fn`s, as public items are part of the API and we don't want to have clippy suggest breaking changes.

**Comment by matthiaskrgr** [comment]
Duplicate of https://github.com/rust-lang/rust-clippy/issues/5969

**Comment by emilk** [comment]
@matthiaskrgr thanks - I searched for "unnecessary result" and "needless result", but not "unneeded result" 🤦‍♂️ 

Anyway, very happy to see this being worked on in https://github.com/rust-lang/rust-clippy/pull/6070 :)
